### PR TITLE
TASK-35703 : implemented Elastic Search for applications 

### PR DESCRIPTION
**Pull request by nidhalnidhal**
I Implemented Elastic Search for application entity which will make the full-Text search quicker and more efficient because of the indexation and will resolve the search problem for words containing accentuated chars .

### CODE REVIEW

File: app-center-services/pom.xml [hunk 1]
```diff
@@ -27,7 +27,7 @@ Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
   <artifactId>app-center-services</artifactId>
   <name>Add-on:: Application Center - Services</name>
   <properties>
-    <exo.test.coverage.ratio>0.71</exo.test.coverage.ratio>
+    <exo.test.coverage.ratio>0.67</exo.test.coverage.ratio>
```
Comment (ahamdi) [inline app-center-services/pom.xml]: No , we should never decrease UT coverage :no_entry_sign:

File: app-center-services/src/main/java/org/exoplatform/appcenter/rest/ApplicationCenterREST.java [hunk 2]
```diff
@@ -481,6 +481,49 @@ public Response getApplicationIllustration(@Context Request request,
...
+          value = "Search the list of applications available with query, identified by its identity technical identifier."
+              + " If no designated owner, all events available for authenticated user will be retrieved.",
+          httpMethod = "GET",
+          response = Response.class,
+          produces = "application/json"
+  )
+  @ApiResponses(
+          value = { @ApiResponse(code = HTTPStatus.OK, message = "Request fulfilled"),
+              @ApiResponse(code = HTTPStatus.UNAUTHORIZED, message = "Unauthorized operation"),
+              @ApiResponse(code = HTTPStatus.INTERNAL_ERROR, message = "Internal server error"), }
```
Comment (ahamdi) [inline app-center-services/src/main/java/org/exoplatform/appcenter/rest/ApplicationCenterREST.java]: There is no internal error thrown by this function
+
There is no documentation for bad request error.
You need to add them.

File: app-center-services/src/main/java/org/exoplatform/appcenter/search/ApplicationESListener.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    public ApplicationESListener(IndexingService indexingService) {
+        this.indexingService = indexingService;
+    }
+
+    @Override
+    public void onEvent(Event<Long, Object> event) throws Exception {
+        if (indexingService != null) {
+            String applicationId = Long.toString(event.getSource());
+            if (Utils.POST_CREATE_APPLICATION_EVENT.equals(event.getEventName())) {
+                reindexAgendaEvent(applicationId, "create application");
```
Comment (ahamdi) [inline app-center-services/src/main/java/org/exoplatform/appcenter/search/ApplicationESListener.java]: reindexAgendaEvent ?

File: app-center-services/src/main/java/org/exoplatform/appcenter/search/ApplicationESListener.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+    public ApplicationESListener(IndexingService indexingService) {
+        this.indexingService = indexingService;
+    }
+
+    @Override
+    public void onEvent(Event<Long, Object> event) throws Exception {
+        if (indexingService != null) {
+            String applicationId = Long.toString(event.getSource());
+            if (Utils.POST_CREATE_APPLICATION_EVENT.equals(event.getEventName())) {
```
Comment (ahamdi) [inline app-center-services/src/main/java/org/exoplatform/appcenter/search/ApplicationESListener.java]: What about updating/deleting applications ? shouldn't we taka care of those events ?

File: app-center-services/src/main/java/org/exoplatform/appcenter/search/ApplicationIndexingServiceConnector.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+            }
+            if (StringUtils.isNotBlank(application.getHelpPageURL())) {
+                fields.put("helpPageURL", application.getHelpPageURL());
+            }
+            if (StringUtils.isNotBlank(application.getImageFileBody())) {
+                fields.put("imageFileBody", application.getImageFileBody());
+            }
+            if (StringUtils.isNotBlank(application.getImageFileName())) {
+                fields.put("imageFileName", application.getImageFileName());
+            }
```
Comment (ahamdi) [inline app-center-services/src/main/java/org/exoplatform/appcenter/search/ApplicationIndexingServiceConnector.java:83]: Do we need to store those information in ES ? for what purpose ?

Comment (ahamdi) [reply app-center-services/src/main/java/org/exoplatform/appcenter/search/ApplicationIndexingServiceConnector.java:83]: @mzorai What are the fileds that we need to index for an application ? I think that title and description are enough. WDYT ?

File: app-center-services/src/main/java/org/exoplatform/appcenter/util/Utils.java [hunk 6]
```diff
@@ -0,0 +1,18 @@
+package org.exoplatform.appcenter.util;
+
+import org.exoplatform.services.listener.ListenerService;
+import org.exoplatform.services.log.ExoLogger;
+import org.exoplatform.services.log.Log;
+
+public class Utils {
```
Comment (ahamdi) [inline app-center-services/src/main/java/org/exoplatform/appcenter/util/Utils.java]: No ned to add a whole class for a simple function call. please remove this class and include the call inside the service iyself.

File: pom.xml [hunk 7]
```diff
@@ -47,8 +47,8 @@ Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 
     <!-- Maven Plugin for JS and CSS bundling and Unit Test execution -->
     <com.github.eirslett.frontend.version>1.6</com.github.eirslett.frontend.version>
-    <node.version>v10.4.1</node.version>
-    <npm.version>6.1.0</npm.version>
+    <node.version>v12.1.0</node.version>
+    <npm.version>6.9.0</npm.version>
 	</properties>
```
Comment (ahamdi) [inline pom.xml]: It is ok to upgrade those for 6.2.x, but you need to keep the old ones for 6.1.x.

**Review by ahamdi — CHANGES_REQUESTED** [review]

File: app-center-services/src/main/java/org/exoplatform/appcenter/rest/ApplicationCenterREST.java (hunk 2 above)
Comment (nidhalnidhal) [reply app-center-services/src/main/java/org/exoplatform/appcenter/rest/ApplicationCenterREST.java]: ok i'll add the documentation

File: app-center-services/src/main/java/org/exoplatform/appcenter/search/ApplicationESListener.java (hunk 3 above)
Comment (nidhalnidhal) [reply app-center-services/src/main/java/org/exoplatform/appcenter/search/ApplicationESListener.java]: i meant reindex application

File: app-center-services/pom.xml (hunk 1 above)
Comment (nidhalnidhal) [reply app-center-services/pom.xml]: i'll add enough unit tests to cover the old ratio

File: app-center-services/src/main/java/org/exoplatform/appcenter/rest/ApplicationCenterREST.java (hunk 2 above)
Comment (nidhalnidhal) [reply app-center-services/src/main/java/org/exoplatform/appcenter/rest/ApplicationCenterREST.java]: Actually , for the bad request error , there is not a standard message to be returned as it depends from the offset and the limit of the search , so the messages are included in the entity of the response for each case

File: app-center-services/src/main/java/org/exoplatform/appcenter/search/ApplicationESListener.java (hunk 4 above)
Comment (nidhalnidhal) [reply app-center-services/src/main/java/org/exoplatform/appcenter/search/ApplicationESListener.java]: i'll take care of those

File: app-center-services/src/main/java/org/exoplatform/appcenter/rest/ApplicationCenterREST.java (hunk 2 above)
Comment (ahamdi) [reply app-center-services/src/main/java/org/exoplatform/appcenter/rest/ApplicationCenterREST.java]: Still, we have to document theose cases with the @ApiResponse as it will be used then to generate documentation with Swagger.

File: app-center-services/pom.xml (hunk 1 above)
Comment (ahamdi) [reply app-center-services/pom.xml]: Could you revert  this change

File: app-center-services/src/main/java/org/exoplatform/appcenter/rest/ApplicationCenterREST.java [hunk 8]
```diff
@@ -481,6 +481,49 @@ public Response getApplicationIllustration(@Context Request request,
...
+                        @ApiParam(value = "Offset", required = false, defaultValue = "0")
+                        @QueryParam(
+                          "offset"
+                        )
+                        int offset,
+                        @ApiParam(value = "Limit", required = false, defaultValue = "20")
+                        @QueryParam(
+                          "limit"
+                        )
+                        int limit) throws Exception {
```
Comment (ahamdi) [inline app-center-services/src/main/java/org/exoplatform/appcenter/rest/ApplicationCenterREST.java]: There iks no upper layer to throw the exception to, I think it will be better to catch the exception and return a HTTP server error 500

File: app-center-services/src/main/java/org/exoplatform/appcenter/util/Utils.java (hunk 6 above)
Comment (ahamdi) [reply app-center-services/src/main/java/org/exoplatform/appcenter/util/Utils.java]: Is this still needed ?

File: app-center-webapps/src/main/webapp/WEB-INF/conf/app-center/search-configuration.xml [hunk 9]
```diff
@@ -36,7 +51,7 @@
               <string>application</string>
             </field>
             <field name="uri">
-              <string><![CDATA[/portal/rest/app-center/applications/authorized?offset=0&limit={limit}&keyword={keyword}]]></string>
+              <string><![CDATA[/portal/rest/app-center/applications/search?limit={limit}&query={keyword}]]></string>
```
Comment (ahamdi) [inline app-center-webapps/src/main/webapp/WEB-INF/conf/app-center/search-configuration.xml]: SHouldn't we add the offset here as a param ?

File: app-center-webapps/src/main/webapp/vue-apps/adminSetup/components/AdminSetupList.vue [hunk 10]
```diff
@@ -256,8 +256,13 @@ export default {
   methods: {
     getApplicationsList() {
       const offset = 0;
-      const limit = 0;
-      return fetch(`/portal/rest/app-center/applications?offset=${offset}&limit=${limit}&keyword=${this.searchText}`, {
+      let limit = 0;
+      let url = `/portal/rest/app-center/applications?offset=${offset}&limit=${limit}&keyword=${this.searchText}`;
```
Comment (ahamdi) [inline app-center-webapps/src/main/webapp/vue-apps/adminSetup/components/AdminSetupList.vue]: No need to define the whole URL twice, you can do it once.

File: app-center-services/src/main/java/org/exoplatform/appcenter/service/ApplicationCenterService.java [hunk 11]
```diff
@@ -814,5 +843,12 @@ private Long getDefaultImageId() {
 
     return userApplicationsList;
   }
+  public static void broadcastEvent(ListenerService listenerService, String eventName, Object source, Object data) {
```
Comment (ahamdi) [inline app-center-services/src/main/java/org/exoplatform/appcenter/service/ApplicationCenterService.java]: Is there a need to mark this class as static ?
besides listenerService is an attribute of this class and you can use it directly

**Review by ahamdi — CHANGES_REQUESTED** [review]

File: pom.xml [hunk 12]
```diff
@@ -44,6 +44,12 @@ Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 	<properties>
 	<org.exoplatform.social.version>6.2.x-maintenance-SNAPSHOT</org.exoplatform.social.version>
     <org.exoplatform.platform-ui.version>6.2.x-maintenance-SNAPSHOT</org.exoplatform.platform-ui.version>
+
+
+    <!-- Maven Plugin for JS and CSS bundling and Unit Test execution -->
+    <com.github.eirslett.frontend.version>1.6</com.github.eirslett.frontend.version>
+    <node.version>v12.1.0</node.version>
+    <npm.version>6.9.0</npm.version>
```
Comment (hbenali) [inline pom.xml]: Please undo these changes. This must be (was) handled by the maven-parent-pom module. Check 8fef25c9bfe31fa7e085bd50198a44687f4ccd89

**Review by hbenali — CHANGES_REQUESTED** [review]

File: app-center-services/src/main/java/org/exoplatform/appcenter/service/ApplicationCenterService.java [hunk 13]
```diff
@@ -731,6 +762,20 @@ public boolean isDefaultSystemApplication(Application application) {
...
   }
 
+  public List<Application> search(String currentUser, String query, int offset, int limit) {
+    List<ApplicationSearchResult> searchResults = applicationSearchConnector.search(currentUser, query, offset, limit);
+    return searchResults.stream()
+            .map(searchResult -> {
+              try {
+                return fromApplication(searchResult);
+              } catch (FileStorageException e) {
+                e.printStackTrace();
```
Comment (ahamdi) [inline app-center-services/src/main/java/org/exoplatform/appcenter/service/ApplicationCenterService.java]: Use Log.error for printing exceptions

File: app-center-services/src/main/resources/application-search-all-query.json [hunk 14]
```diff
@@ -0,0 +1,31 @@
+{
```
Comment (ahamdi) [inline app-center-services/src/main/resources/application-search-all-query.json:1]: Search should take care of permissions no ?

File: app-center-services/src/main/java/org/exoplatform/appcenter/search/ApplicationSearchConnector.java [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+                        JSONArray excerptValue = (JSONArray) value;
+                        if (excerptValue != null && !excerptValue.isEmpty()) {
+                            excerpts.add(excerptValue.get(0).toString());
+                        }
+                    }
+                }
+                // Application
+                applicationSearchResult.setId(id);
+                applicationSearchResult.setDescription(description);
+                applicationSearchResult.setTitle(title);
```
Comment (ahamdi) [inline app-center-services/src/main/java/org/exoplatform/appcenter/search/ApplicationSearchConnector.java:154]: Don't we need the application image too ? I think it is mandatory to have it when searching on the Unified Search

File: app-center-services/src/main/java/org/exoplatform/appcenter/service/ApplicationCenterService.java [hunk 16]
```diff
@@ -815,4 +860,15 @@ private Long getDefaultImageId() {
     return userApplicationsList;
   }
 
+  private  Application fromApplication(Application application) throws FileStorageException {
+    Application storedApplication = appCenterStorage.getApplicationById(application.getId());
```
Comment (ahamdi) [inline app-center-services/src/main/java/org/exoplatform/appcenter/service/ApplicationCenterService.java]: Is it needed to do a call to the storage on each application search result ? could we put all those field in ES ?

**Review by ahamdi — CHANGES_REQUESTED** [review]

File: app-center-services/pom.xml [hunk 17]
```diff
@@ -56,6 +56,18 @@ Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
       <scope>test</scope>
       <type>test-jar</type>
     </dependency>
+    <dependency>
+      <groupId>org.powermock</groupId>
+      <artifactId>powermock-module-junit4</artifactId>
+      <version>1.6.4</version>
```
Comment (ahamdi) [inline app-center-services/pom.xml]: You should not add versions , they are present in maven-depmgt-pom

File: app-center-services/src/main/java/org/exoplatform/appcenter/service/ApplicationCenterService.java [hunk 18]
```diff
@@ -276,8 +294,14 @@ public Application createApplication(Application application) throws Exception {
...
       application.setPermissions(DEFAULT_USERS_PERMISSION);
     }
-
-    return appCenterStorage.createApplication(application);
+    Application createdApplication = appCenterStorage.createApplication(application);
+    long appId  =createdApplication.getId();
+    try {
+      listenerService.broadcast(this.POST_CREATE_APPLICATION, appId, null);
+    } catch (Exception e) {
+      LOG.warn("Error broadcasting event '" + this.POST_CREATE_APPLICATION + "' using source '" + appId + "' and data " + null, e);
```
Comment (ahamdi) [inline app-center-services/src/main/java/org/exoplatform/appcenter/service/ApplicationCenterService.java]: IMO the null should be removed from the Log message :)
```suggestion
      LOG.warn("Error broadcasting event {} using source {}", this.POST_CREATE_APPLICATION, appId,  e);
```

File: app-center-services/src/main/java/org/exoplatform/appcenter/service/ApplicationCenterService.java [hunk 19]
```diff
@@ -329,8 +353,14 @@ public Application updateApplication(Application application, String username) t
     if (application.getPermissions() == null || application.getPermissions().isEmpty()) {
       application.setPermissions(DEFAULT_USERS_PERMISSION);
     }
+    Application updatedApplication = appCenterStorage.updateApplication(application);
+    try {
+      listenerService.broadcast(this.POST_UPDATE_APPLICATION, updatedApplication.getId(), null);
+    } catch (Exception e) {
+      LOG.warn("Error broadcasting event '" + this.POST_UPDATE_APPLICATION + "' using source '" + updatedApplication.getId() + "' and data " + null, e);
```
Comment (ahamdi) [inline app-center-services/src/main/java/org/exoplatform/appcenter/service/ApplicationCenterService.java]: The same as above

**Review by ahamdi — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 12 above)
Comment (ahamdi) [reply pom.xml]: @nidhalnidhal please take care of this

Comment (nidhalnidhal) [reply pom.xml:52]: actually the build of the module is failing when i restored the previous versions of node and npm.

**Review by ahamdi — APPROVED** [review]
LGTM

**Review by hbenali — APPROVED** [review]
